server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java

```diff
@@ -968,8 +977,10 @@ public void test_copy_from_with_fail_fast_property_can_kill_all_nodes_with_failu
             nodeIds.add((String) response.rows()[i][1]);
         }
 
-        for (int i = 0; i < nodeIds.size(); i++) {
+        long accumulateFailingUriCount = 0;
+        long accumulatePassingUriCount = 0;
 
+        for (int i = 0; i < nodeIds.size(); i++) {
```

Comment:
I don't completely understand this test scenario.
Afaik we want to test that the failing record is once consumed on the handler node and once on a non-handler node.
So wouldn't this be much simple if we create the dataset upfront with a concrete node/readerid and then execute the `COPY FROM` twice, once on node which consumes the error row (thus failure happen on the handler node) and once on the node which won't concume and error prone rows (thus failure will happen on a non-handler node).
Wouldn't this also make the scenario much more deterministic and safe?
As far as I understood we must somehow ensure that all nodes will still be processing rows when the error is happening. Thus if the error happens on the first row, the error propagation must be faster than consuming x rows.
But I wonder that we must create more rows to make it succeed on slower machines, I'd expect the opposite.

Reply:
Thank you @seut, when I first noticed that the `x` rows are entirely consumed before the error propagation, my instinct was to increase `x`. Today I spent more time experimenting with reduced `x`, and realized that  increasing/decreasing `x` affects the mean of multiple trials of copy stmt not the variance. Now, I am trying with low `x` and repeated with a certain number of `maxTry` like my other tests to ignore the occasions with high variance.

Also, I agree to
> Wouldn't this also make the scenario much more deterministic and safe?

But, how could I find out which node is the handler node?

Reply:
> But, how could I find out which node is the handler node?

It's about how to ensure that the statement will be send to a concrete node as the node receiving the statement is the handler node.
See e.g. https://github.com/crate/crate/blob/master/server/src/test/java/io/crate/integrationtests/ReadOnlyNodeIntegrationTest.java how this can be ensured.

So we'd create a dataset which ensures that e.g. the node stored inside `internalCluster().getNodeNames()[0]` will process the failing row. Then we'll issue the statement once against this node (error happens on the "handler" node) and once against the other node (error happens not on the "handler" node).